src/bin/nydus-image/builder/directory.rs

```diff
@@ -147,39 +149,57 @@ impl Builder for DirectoryBuilder {
         let mut blob_writer = if let Some(blob_stor) = ctx.blob_storage.clone() {
             Some(ArtifactWriter::new(blob_stor, ctx.inline_bootstrap)?)
         } else {
-            None
+            return Err(anyhow!(
+                "the target blob path should always be valid for directory builder"
+            ));
         };
 
         // Dump blob file
-        let mut blob = Blob::new();
         timing_tracer!(
-            { blob.dump(ctx, &mut bootstrap_ctx.nodes, blob_mgr, &mut blob_writer) },
+            { Blob::dump(ctx, &mut bootstrap_ctx.nodes, blob_mgr, &mut blob_writer) },
             "dump_blob"
         )?;
 
+        // Dump bootstrap file
+        let blob_table = blob_mgr.to_blob_table(ctx)?;
+        bootstrap.dump(
+            ctx,
+            &mut bootstrap_mgr.bootstrap_storage,
+            &mut bootstrap_ctx,
+            &blob_table,
+        )?;
+
         if let Some((_, blob_ctx)) = blob_mgr.get_current_blob() {
```

Comment:
If no blob is generated, we still need to write bootstrap data and tar header (`TAR_BOOTSTRAP_NAME`) into the blob if `inline-bootstrap` is enabled.

Reply:
Fixed